test/should_fail/nonempty_list_match_in_head_nonexhaustive.erl

```diff
@@ -0,0 +1,15 @@
+-module(nonempty_list_match_in_head_nonexhaustive).
+
+-export([f/1]).
+
+%% For the infinite loop bug to happen all the conditions must be met:
+%% - the function matches on a non-empty list
+%% - the type is a union of something and a list of itself - if it's not a union,
+%%   typechecker doesn't loop
+%% - the type is recursive - a list of something else doesn't trigger the infinite loop
+%% - the pattern is bound to a variable (doesn't matter if used or unused)
```

Comment:
Interesting.

Since you found out that also maps are normalized inside their elements, is there a similar case for maps? `t() :: {} | #{t() => t()}`?

Btw, I think an issue is a better place for such detailed bug information. Especially when the bug is fixed, such description makes less sense in the source code. A link to the issue is enough as motivation for the test.